pkg/cli/sql_util.go

```diff
@@ -71,10 +80,143 @@ func (c *sqlConn) ensureConn() error {
+		if err := c.checkServerMetadata(); err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
+var clientVersionShort = build.GetInfo().Short()
+var unknownServerBuild = build.Info{
+	GoVersion:    "?",
+	Tag:          "?",
+	Time:         "?",
+	Revision:     "?",
+	CgoCompiler:  "?",
+	Platform:     "?",
+	Distribution: "?",
+	Type:         "?",
+	Dependencies: nil,
+}
+
+// checkServerMetadata reports the server version, cluster ID and the
+// node ID upon the initial connection or if either has changed since
+// the last connection, based on the last known values in the sqlConn
+// struct.
+func (c *sqlConn) checkServerMetadata() error {
+	if !isInteractive {
+		// Version reporting is just noise in non-interactive sessions.
+		return nil
+	}
+
+	newNodeID := ""
+	newServerBuildInfo := unknownServerBuild
+	newClusterID := ""
+
+	// Retrieve the cluster ID.
+	clusterVal, _ := c.getServerValue("cluster ID", "SELECT crdb_internal.cluster_id()::STRING")
+	if s, ok := clusterVal.(string); ok {
+		newClusterID = s
+	}
+
+	// Retrieve the node ID and server build info.
+	rows, err := c.Query("SELECT * FROM crdb_internal.node_build_info", nil)
+	if err == driver.ErrBadConn {
+		return err
+	}
+	if err != nil {
+		fmt.Fprintln(stderr, "unable to retrieve the server's version")
+	} else {
+		defer func() { _ = rows.Close() }()
+
+		// Read the node_build_info table as an array of strings.
+		rowVals, err := getAllRowStrings(rows, true /* showMoreChars */)
+		if err != nil || len(rowVals) == 0 || len(rowVals[0]) != 3 {
+			fmt.Fprintln(stderr, "error while rerieving the server's version")
+			return err
+		}
+
+		// Extract the build.Info fields from the query results.
+		var zeroValue reflect.Value
+		infoStruct := reflect.ValueOf(&newServerBuildInfo).Elem()
+		for _, row := range rowVals {
+			newNodeID = row[0]
+			if vField := infoStruct.FieldByName(row[1]); vField != zeroValue {
+				vField.Set(reflect.ValueOf(row[2]))
+			}
+		}
+	}
+
+	// Report the server version only if it the revision has been
+	// fetched successfully, and the revision has changed since the last
+	// connection.
+	if newServerBuildInfo != c.serverBuildInfo && newServerBuildInfo.Revision != "?" {
+		c.serverBuildInfo = newServerBuildInfo
+
+		serverVersionShort := newServerBuildInfo.Short()
+		isSame := ""
+		if serverVersionShort == clientVersionShort {
+			isSame = " (same version as client)"
+		}
+
+		if isSame == "" {
+			fmt.Println("# Client version:", clientVersionShort)
+		}
+		fmt.Printf("# Server version: %s%s\n", serverVersionShort, isSame)
+	}
+
+	// Report the cluster ID only if it it could be fetched
+	// successfully, and it has changed since the last connection.
+	if newClusterID != c.clusterID && newClusterID != "" {
+		// As a special case, we complain loudly if the cluster ID
+		// actually changes. This is sign of a potentially very bad load
+		// balancer configuration.
+		if c.clusterID != "" {
+			fmt.Print("\n# WARNING: the cluster ID has changed!\n\n")
```

Comment:
This isn't enough, we should error out if reconnecting to a different cluster. I'd even be happy with a hard `log.Fatal`